File: Computational_Physics_2/1.Lectures/7.Computacion_en_parallelo_MPI/examples/examples_fortran/example_6_Reduce.f90
PROGRAM main
   USE mpi
   IMPLICIT NONE
   INTEGER, parameter :: N=1001
   INTEGER :: ierr, i, rank, nprocs
   INTEGER :: nstart, nstop, npart, ncount, nrem
   DOUBLE PRECISION, allocatable :: vec(:)
   DOUBLE PRECISION :: local_sum, total_sum

   CALL MPI_INIT(ierr)
   CALL MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
   CALL MPI_Comm_size(MPI_COMM_WORLD, nprocs, ierr)

   ncount = N/nprocs
   nrem = MOD(N,nprocs)

   IF (rank < nrem) THEN
      nstart = rank * (ncount + 1)
      nstop = nstart + ncount
   ELSE
     nstart = rank * ncount + nrem
     nstop = nstart + (ncount - 1)
   END IF

   npart = nstop-nstart+1

   allocate(vec(npart))
   
   DO i=1,npart
     vec(i) = 1.0D0
   END DO
 
   local_sum = 0.0D0

   DO i=1,npart
      local_sum = local_sum + vec(i)
   END DO

   WRITE(*,’(A,I3,A,F15.8)’) "[",rank,"] part : ",local_sum

   CALL MPI_BARRIER(MPI_COMM_WORLD,ierr)
   CALL MPI_REDUCE(local_sum, total_sum, 1, MPI_DOUBLE, MPI_SUM, 0 , MPI_COMM_WORLD, ierr)

   IF (rank.eq.0) THEN
      WRITE(*,*) "Sum : ", total_sum
   END IF

   deallocate(vec)

   CALL MPI_FINALIZE(ierr)

END PROGRAM

File: Computational_Physics_2/1.Lectures/7.Computacion_en_parallelo_MPI/examples/examples_C/example_6_Reduce.c
/*
https://indico.in2p3.fr/event/20625/contributions/101997/attachments/67854/95270/mpi.pdf
*/
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
  int i, N;
  int nprocs, rank;

  MPI_Init(&argc,&argv);
  MPI_Comm_size(MPI_COMM_WORLD,&nprocs);
  MPI_Comm_rank(MPI_COMM_WORLD,&rank);
  
  double *vec;
  double local_sum,sum;

  N=1001;   

  int count = N/nprocs;
  int remainder = N%nprocs;
  int start, stop;

  if (rank < remainder) {
     start = rank * (count + 1);
     stop = start + count;
  } else {
     start = rank * count + remainder;
     stop = start + (count - 1);
  }

  int npart = stop-start+1;
  vec = malloc(sizeof(double)*npart);
  
  for (i=0; i<npart; i++) {
    vec[i] = 1.0;
  }

  local_sum = 0.0;

  for (i=0; i<npart; i++) {
     local_sum += vec[i];
  }

  printf("[ %d ] part : %g\n",rank,local_sum);

  MPI_Barrier(MPI_COMM_WORLD);
  MPI_Reduce(&local_sum,&sum,1,MPI_DOUBLE,MPI_SUM,0,MPI_COMM_WORLD);
 
  if (rank == 0) {
    printf("Sum : %g\n",sum);
  }

  MPI_Finalize();
  return 0;
}
